repairnator/src/main/java/fr/inria/spirals/repairnator/process/inspectors/ProjectInspector4Bears.java

```diff
@@ -1,124 +1,83 @@
-                AbstractStep testProject = new TestProject(this);
-                this.testInformations = new GatherTestInformation(this, new BuildShouldPass());
-
-                AbstractStep checkoutPreviousBuild = new CheckoutPreviousBuild(this);
-                AbstractStep buildRepoForPreviousBuild = new BuildProject(this);
-                AbstractStep testProjectForPreviousBuild = new TestProject(this);
-                if (this.getPreviousBuild().getBuildStatus() == BuildStatus.FAILED) {
-                    AbstractStep gatherTestInformation = new GatherTestInformation(this, new BuildShouldFail());
-
-                    cloneRepo.setNextStep(buildRepo).setNextStep(testProject).setNextStep(this.testInformations)
-                            .setNextStep(checkoutPreviousBuild).setNextStep(buildRepoForPreviousBuild)
-                            .setNextStep(testProjectForPreviousBuild).setNextStep(gatherTestInformation);
-
-                    lastStep = gatherTestInformation;
-                } else {
-                    AbstractStep gatherTestInformation = new GatherTestInformation(this, new BuildShouldPass());
-                    AbstractStep checkoutSourceCodeForPreviousBuild = new CheckoutSourceCodeForPreviousBuild(this);
-                    AbstractStep buildRepoForPreviousBuild2 = new BuildProject(this);
-                    AbstractStep testProjectForPreviousBuild2 = new TestProject(this);
-                    AbstractStep gatherTestInformation2 = new GatherTestInformation(this, new BuildShouldFail());
-
-                    cloneRepo.setNextStep(buildRepo).setNextStep(testProject).setNextStep(this.testInformations)
-                            .setNextStep(checkoutPreviousBuild).setNextStep(buildRepoForPreviousBuild)
-                            .setNextStep(testProjectForPreviousBuild).setNextStep(gatherTestInformation)
-                            .setNextStep(checkoutSourceCodeForPreviousBuild).setNextStep(buildRepoForPreviousBuild2)
-                            .setNextStep(testProjectForPreviousBuild2).setNextStep(gatherTestInformation2);
-
-                    lastStep = gatherTestInformation2;
-                }
+        AbstractStep firstStep = null;
+        AbstractStep lastStep = null;
+
+        // Clone, build, test and gather test information for the
+        // current passing build to ensure it is reproducible
+        AbstractStep cloneRepo = new CloneRepository(this);
+        AbstractStep buildRepo = new BuildProject(this);
+        AbstractStep testProject = new TestProject(this);
+        this.testInformations = new GatherTestInformation(this, new BuildShouldPass());
+
+        AbstractStep checkoutPreviousBuild = new CheckoutPreviousBuild(this);
+        AbstractStep buildRepoForPreviousBuild = new BuildProject(this);
+        AbstractStep testProjectForPreviousBuild = new TestProject(this);
+        if (this.getPreviousBuild().getBuildStatus() == BuildStatus.FAILED) {
```

Comment:
Hmmm in the case of passing/passing ok, but here it's the specific "passing/fail" so it could be tested with the PASSING_AND_FAIL status you should add ^^